### PR TITLE
Fix empty `host.id`

### LINKED ISSUE #4312: `host.id` is set to empty value in containerized setups
### Description

The `[resource.WithHostID`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk@v1.16.0/resource#WithHostID) option sets the `host.id` resource attribute to the empty string on (certain) containerized setups. The attribute should be set in accordance to [the specification](https://github.com/open-telemetry/semantic-conventions/blob/c6efa4b2b44f9950b332a46311834e23b85bf979/docs/resource/host.md) which states that `host.id` should be a:

> Unique host ID. For Cloud, this must be the instance_id assigned by the cloud provider. For non-containerized systems, this should be the machine-id. See the table below for the sources to use to determine the machine-id based on operating system.

An empty string is not a unique host ID.

### Environment

- OS: Ubuntu 22.04.2, likely irrelevant
- Architecture: x86_64, likely irrelevant
- Go Version: go1.19, likely irrelevant
- opentelemetry-go version: Tested with v1.16.0

### Steps To Reproduce

Run the binary resulting from the following program:

```go
package main

import (
	"context"
	"fmt"

	"go.opentelemetry.io/otel/sdk/resource"
)

func main() {
	res, err := resource.New(context.Background(),
		resource.WithHostID(),
	)

	if err != nil {
		panic(err)
	}

	fmt.Println(res)
}
```

On any of the following images:
- `alpine`
- `ubuntu`
- `centos`
- a `FROM scratch` image (see open-telemetry/opentelemetry-collector-contrib/issues/24230)

The resulting output is:

```
host.id=
```

### Expected behavior

The `host.id` resource attribute is either not set or set to a value that is specification compliant.

----

cc @mwear. I have not tried to reproduce this on opentelemetry-js but it's likely the issue occurs there as well.

### PR DESCRIPTION
**Pull request by mwear**
Fixes #4312. 

When reading a non-existent file, `readFile` was returning `"", nil` instead of `"", err` resulting in an empty host id. The original intent was to return an error in this case, but this will make the example in #4312 (pasted below) panic:

```go
package main

import (
	"context"
	"fmt"

	"go.opentelemetry.io/otel/sdk/resource"
)

func main() {
	res, err := resource.New(context.Background(),
		resource.WithHostID(),
	)

	if err != nil {
		panic(err)
	}

	fmt.Println(res)
}
```

In a containerized environment, it's very likely that HostID detection will fail. That calls into question how we should handle this case. Should the `hostIDDetector` ignore errors and return an empty resource, or should it propagate the error (as per the original intent)? There is a similar detector in JS, but they work a little differently than the ones in the Go SDK. The [HostDetector](https://github.com/open-telemetry/opentelemetry-js/blob/87f21ef8aecaa1e52ff9200a99497276ffa2956b/packages/opentelemetry-resources/src/platform/node/HostDetectorSync.ts#L40-L46) detects all `host.*` attributes. It will omit `host.id` if it's not found, but return the other found attributes in a new resource. I think the Go SDK equivalent would be to make the `hostIDDetector` a best-effort detector, and return an empty resource if it fails. Do any other detectors do this? Does anyone have thoughts or opinions? (cc: @mx-psi)

### CODE REVIEW

**Review by MrAlias — COMMENTED** [review]
Should there be a test for this?

**Review by pellared — APPROVED** [review]

**Review by mx-psi — APPROVED** [review]
I think this is a net improvement, we can continue discussing what the behavior should actually be in a containerized setup separately.

To me silently failing sounds like "there is no valid `host.id` on this setup". This is not necessarily the case if we fail to read the files specified on the spec, since `host.id` could also be a cloud provider ID (e.g. AWS EC2 instance id). But I don't have context on what other detectors do to have a strong opinion here.

**Comment by pellared** [comment]
> I think this is a net improvement, we can continue discussing what the behavior should actually be in a containerized setup separately.
> 
> To me silently failing sounds like "there is no valid `host.id` on this setup". This is not necessarily the case if we fail to read the files specified on the spec, since `host.id` could also be a cloud provider ID (e.g. AWS EC2 instance id). But I don't have context on what other detectors do to have a strong opinion here.

This PR fixes the bug. There should be a new issue (enhancement) that should be about providing a `host.id` value for the mentioned use cases.

**Comment by mwear** [comment]
> Should there be a test for this?

`readFile` ends up getting mocked in all of the higher level tests. I can specifically add tests for the `readFile` function. Let me know if you'd like me to go that route.

**Comment by mwear** [comment]
Since `readFile` does not get exercised in normal tests due to mocking, I went ahead and added a couple of tests for it.

**Review by mx-psi — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -35,6 +35,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 - Correctly format log messages from the `go.opentelemetry.io/otel/exporters/zipkin` exporter. (#4143)
 - Log an error for calls to `NewView` in `go.opentelemetry.io/otel/sdk/metric` that have empty criteria. (#4307)
+- Fix `readFile` in `go.opentelemetry.io/otel/sdk/resource` to return an error for non-existent files. `resource.WithHostID()` would previously return a resource with an empty `host.id` instead of returning an error. (#4317)
```
[the remark this answers is not in the document]
Comment (pellared) [reply CHANGELOG.md:38]: Maybe  simply

```suggestion
- Fix `resource.WithHostID()` to not set an empty `host.id`. (#4317)
```

**Review by pellared — APPROVED** [review]

**Review by dmathieu — APPROVED** [review]

**Pull request merged**